### PR TITLE
Transition to new schema format

### PR DESCRIPTION
**Pull request by ThomasColthurst**
As proposed in the discussion of https://github.com/probcomp/hierarchical-irm/issues/53

### CODE REVIEW

File: cxx/util_distribution_variant.cc [hunk 1]
```diff
@@ -15,7 +15,12 @@
 #include "distributions/stringcat.hh"
 #include "util_observation.hh"
 
-DistributionSpec::DistributionSpec(const std::string& dist_str) {
+DistributionSpec::DistributionSpec(
+    const std::string& dist_str,
+    const std::map<std::string, std::string>& _distribution_args):
+  distribution_args(_distribution_args) {
+  printf("DEBUG:  dist_str = %s, dist_args.size = %ld\n",
```
Comment (emilyfertig) [inline cxx/util_distribution_variant.cc:22]: Remove print statements? (+ below)

File: cxx/util_io.cc [hunk 2]
```diff
@@ -11,36 +12,166 @@
...
+              assert(base_domains[i] == trel.domains[i] &&
+                     "The first domains of a noisy relation must be the same "
+                     "as the domains of the base relation.");
+            }
+          }
+        }, trelation);
+  }
+}
+
 T_schema load_schema(const std::string& path) {
```
Comment (emilyfertig) [inline cxx/util_io.cc:34]: Could you add the new test in `util_io_test` from #95 that tests the `with_noisy` schema?

File: cxx/util_io.cc [hunk 3]
```diff
@@ -11,36 +12,166 @@
...
+             tokens[2].val.c_str(), (int)(tokens[2].type));
+      assert(false);
+    }
+    std::string dist_name = tokens[2].val;
+
+    size_t i = 3;
+    std::map<std::string, std::string> params;
+    if (tokens[i].val == "[") { // Handle parameters
+      do {
+        i += 1;
```
Comment (emilyfertig) [inline cxx/util_io.cc:83]: nit: increment with `++i`

**Review by emilyfertig — CHANGES_REQUESTED** [review]
Nice, reusing the tokenizer code is much cleaner than what I would have done.

File: cxx/test_schema/with_noisy.schema [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# A schema with noisy relations.
+R1 ~ bernoulli(D1)
+R2 ~ gaussian(R6; D1, D2)
+R3 ~ bigram(D2)
+R4 ~ gaussian(R2; D1, D2, D3)
+R5 ~ sometimes_bitflip(R1; D1)
+R6 ~ normal(R6; D1)
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply cxx/test_schema/with_noisy.schema]: Done.

File: cxx/util_distribution_variant_test.cc [hunk 5]
```diff
@@ -29,12 +29,21 @@ BOOST_AUTO_TEST_CASE(test_distribution_spec) {
   BOOST_TEST((ds.distribution == DistributionEnum::skellam));
   BOOST_TEST(ds.distribution_args.empty());
 
-  DistributionSpec dc = DistributionSpec("categorical(k=6)");
+  //DistributionSpec dc = DistributionSpec("categorical(k=6)");
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply cxx/util_distribution_variant_test.cc]: Done.

File: cxx/util_distribution_variant.cc (hunk 1 above)
Comment (ThomasColthurst) [reply cxx/util_distribution_variant.cc]: Done.

File: cxx/util_io.cc (hunk 2 above)
Comment (ThomasColthurst) [reply cxx/util_io.cc:62]: Done.

File: cxx/util_io.cc (hunk 3 above)
Comment (ThomasColthurst) [reply cxx/util_io.cc]: Done.

**Comment by ThomasColthurst** [comment]
Also cleaned up some of the error checking in load_schema by defining two macros.

File: cxx/util_io.cc [hunk 6]
```diff
@@ -31,6 +31,34 @@ void verify_noisy_relation_domains(const T_schema& schema) {
   }
 }
 
+#define TYPE_CHECK(i, expected_type) \
```
Comment (emilyfertig) [inline cxx/util_io.cc:34]: Why use macros instead of functions here?

File: cxx/util_io.cc [hunk 7]
```diff
@@ -46,56 +74,27 @@ T_schema load_schema(const std::string& path) {
       continue;
     }
 
-    if (tokens.size() < 6) {
-      printf("Error parsing schema line %s:  expected at least six tokens, "
-             "got %ld instead.\n", line.c_str(), tokens.size());
-      assert(false);
-    }
+    size_t i = 0;
+    TYPE_CHECK(i, identifier);
```
Comment (emilyfertig) [inline cxx/util_io.cc:78]: Out of curiosity, where is `identifier` defined? I don't see it.

**Review by emilyfertig — APPROVED** [review]
Thanks!

File: cxx/util_io.cc (hunk 6 above)
Comment (ThomasColthurst) [reply cxx/util_io.cc:34]: So, partly out of habit -- before C++ had closures, macros were a cheap way of not having to list out all of the captured variables.

But macros have a bunch of other handy features that make them good for this sort of error reporting.  You often want to report the line number or function of the code the error happens at for example, and that's easily done within a macros with __LINE__.

Another feature, which I actually use here, is the fact that the input to a macro is treated as text and doesn't have to be a value.  So when I write "TYPE_CHECK(i, identifier)" below, identifier isn't a value, it's just a piece of text.  It becomes a value when TokenType:: is prepended to it, but it is also used textually inside the error reporting string.

File: cxx/util_io.cc (hunk 7 above)
Comment (ThomasColthurst) [reply cxx/util_io.cc:78]: See above explanation.

**Pull request merged**